tests/test_discovery.py

```diff
@@ -413,6 +416,103 @@ def test_userip_missing_is_not_added_to_discovery_uri(self):
       self.assertEqual(e.uri, 'http://example.com')
 
 
+class DiscoveryFromAppEngineCache(unittest.TestCase):
+  def test_appengine_memcache(self):
+    # Hack module import
+    self.orig_import = __import__
+    self.mocked_api = mock.MagicMock()
+
+    def import_mock(name, *args):
+      if name == 'google.appengine.api':
+        return self.mocked_api
+      return self.orig_import(name, *args)
+
+    import_fullname = '__builtin__.__import__'
+    if sys.version_info[0] >= 3:
+      import_fullname = 'builtins.__import__'
+
+    with mock.patch(import_fullname, side_effect=import_mock):
+      ns = 'google-api-client'
+      self.http = HttpMock(datafile('plus.json'), {'status': '200'})
+
+      self.mocked_api.memcache.get.return_value = None
+
+      plus = build('plus', 'v1', http=self.http)
+
+      # memcache.get is called once
+      url = 'https://www.googleapis.com/discovery/v1/apis/plus/v1/rest'
+      self.mocked_api.memcache.get.assert_called_once_with(url, namespace=ns)
+
+      # memcache.set is called once
+      with open(datafile('plus.json')) as f:
+        c = f.read()
+      self.mocked_api.memcache.set.assert_called_once_with(
+        url, c, time=DISCOVERY_DOC_MAX_AGE, namespace=ns)
+
+      # Returns the cached content this time.
+      self.mocked_api.memcache.get.return_value = c
+
+      # Make sure the contents are returned from the cache.
+      # (Otherwise it should through an error)
+      self.http = HttpMock(None, {'status': '200'})
+
+      plus = build('plus', 'v1', http=self.http)
+
+      # memcache.get is called twice
+      self.mocked_api.memcache.get.assert_has_calls(
+        [mock.call(url, namespace=ns), mock.call(url, namespace=ns)])
+
+      # memcahce.set is called just once
+      self.mocked_api.memcache.set.assert_called_once_with(
+        url, c, time=DISCOVERY_DOC_MAX_AGE,namespace=ns)
+
+
+class DictCache(object):
```

Comment:
Why not have this inherit from cache.Cache? That's the interface it implements, right?